cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -301,80 +694,305 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
+		return nil, err
+	}
+	// all done, nothing left to try and mount
+	return nil, nil
+}
+
+func generateMountsModeRecover(mst *initramfsMountsState) error {
+	// steps 1 and 2 are shared with install mode
+	if err := generateMountsCommonInstallRecover(mst); err != nil {
 		return err
 	}
 
-	// 3.1. mount ubuntu-save (if present)
-	haveSave, err := maybeMountSave(disk, boot.InitramfsHostWritableDir, unlockRes.IsDecryptedDevice, nil)
+	// get the disk that we mounted the ubuntu-seed partition from as a
+	// reference point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
 	if err != nil {
 		return err
 	}
 
-	// 3.2 verify that the host ubuntu-data comes from where we expect it to
-	// right device
-	diskOpts := &disks.Options{}
-	if unlockRes.IsDecryptedDevice {
-		// then we need to specify that the data mountpoint is expected to be a
-		// decrypted device, applies to both ubuntu-data and ubuntu-save
-		diskOpts.IsDecryptedDevice = true
-	}
+	// 3. run the state machine logic for mounting partitions, this involves
+	//    trying to unlock then mount ubuntu-data, and then unlocking and
+	//    mounting ubuntu-save
+	//    see the state* functions for details of what each step does and
+	//    possible transition points
+
+	machine, err := func() (machine *stateMachine, err error) {
+		// ensure that the last thing we do after mounting everything is to lock
+		// access to sealed keys
+		defer func() {
+			if err := secbootLockTPMSealedKeys(); err != nil {
+				logger.Noticef("error locking access to sealed keys: %v", err)
+			}
+		}()
 
-	matches, err = disk.MountPointIsFromDisk(boot.InitramfsHostUbuntuDataDir, diskOpts)
+		// first state to execute is to unlock ubuntu-data with the run key
+		machine = newStateMachine(disk)
+		for {
+			final, err := machine.execute()
+			// TODO: consider whether certain errors are fatal or not
+			if err != nil {
+				return nil, err
+			}
+			if final {
+				break
+			}
+		}
+
+		return machine, nil
+	}()
 	if err != nil {
 		return err
 	}
-	if !matches {
-		return fmt.Errorf("cannot validate boot: ubuntu-data mountpoint is expected to be from disk %s but is not", disk.Dev())
-	}
-	if haveSave {
-		// 3.2a we have ubuntu-save, verify it as well
-		matches, err = disk.MountPointIsFromDisk(boot.InitramfsUbuntuSaveDir, diskOpts)
+
+	// 3.1 write out degraded.json if we ended up falling back somewhere
+	if machine.degraded() {
+		b, err := json.Marshal(machine.degradedState)
 		if err != nil {
 			return err
 		}
-		if !matches {
-			return fmt.Errorf("cannot validate boot: ubuntu-save mountpoint is expected to be from disk %s but is not", disk.Dev())
+
+		err = os.MkdirAll(dirs.SnapBootstrapRunDir, 0755)
+		if err != nil {
+			return err
+		}
+
+		// leave the information about degraded state at an ephemeral location
+		err = ioutil.WriteFile(filepath.Join(dirs.SnapBootstrapRunDir, "degraded.json"), b, 0644)
```

Comment:
(nitpick^2) this could be a single line

Reply:
done in https://github.com/snapcore/snapd/pull/9621